File: 033_search_in_rotated_sorted_array/Solution.java
public class Solution {
    public int search(int[] nums, int target) {
        int arrLen = nums.length;
        int left = 0, right = arrLen - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] == target) return mid;
            if (nums[mid] < nums[right]) { // 右侧有序
                if (nums[mid] < target && target <= nums[right]) left = mid + 1;
                else right = mid - 1;
            } else { // 左侧有序
                if (nums[left] <= target && target < nums[mid]) right = mid - 1;
                else left = mid + 1;
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{4, 5, 6, 7, 0, 1, 2};
        Solution solution = new Solution();
        System.out.println(solution.search(nums, 0));
    }
}

File: 033_search_in_rotated_sorted_array/Solution.py
class Solution:
    def search(self, nums: list, target: int) -> int:
        arr_len = len(nums)
        left, right = 0, arr_len - 1

        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target: return mid
            if nums[mid] < nums[right]: # 右侧有序
                if nums[mid] < target <= nums[right]: left = mid + 1
                else: right = mid - 1
            else:
                if nums[left] <= target < nums[mid]: right = mid - 1
                else: left = mid + 1

        return -1


if __name__ == '__main__':
    nums = [4, 5, 6, 7, 0, 1, 2]
    target = 0
    solution = Solution()
    print(solution.search(nums, target))
